### PR TITLE
[Bug]: fix DDPM scheduler arbitrary infer steps count.

### PR DESCRIPTION
**Pull request by dudulightricks**
DDPM allegedly supports evaluation with lower count of steps than in the train -
It allows to set infer timesteps count, and then space them based on the train timesteps
(For example for 1000 train steps and 100 infer steps it make time steps from the form 0, 10, 20, 30...).
Practically, It uses the previous timestep as "t - 1" to calculate the alphas, betas and variance, 
which is incorrect in case of different timesteps count in infer (for example, for 1000 train steps and 100 infer
steps, it should be "t - 10" and not "t - 1").

This is a bug, which cause the result image to be just noise in any case of lower count of timesteps in the infer.
Our fix sets the previous timestep to be the correct one (previous step in the current timesteps array), in those cases, while keeping the backward compatibility.

### CODE REVIEW

**Comment by dudulightricks** [comment]
@patrickvonplaten
@pcuenca 
 Hey, I will be happy to get a review for my suggested changes and I'm not sure who exactly should I ask.
Thanks in advance.

**Comment by dudulightricks** [comment]
@patrickvonplaten
@pcuenca
Hey, I will be happy to get a review for my suggested changes and I'm not sure who exactly should I ask.
Thanks in advance.

File: tests/pipelines/audio_diffusion/test_audio_diffusion.py [hunk 1]
```diff
@@ -118,7 +118,7 @@ def test_audio_diffusion(self):
         assert image.height == self.dummy_unet.sample_size[0] and image.width == self.dummy_unet.sample_size[1]
         image_slice = np.frombuffer(image.tobytes(), dtype="uint8")[:10]
         image_from_tuple_slice = np.frombuffer(image_from_tuple.tobytes(), dtype="uint8")[:10]
-        expected_slice = np.array([255, 255, 255, 0, 181, 0, 124, 0, 15, 255])
+        expected_slice = np.array([69, 255, 255, 255, 0, 0, 77, 181, 12, 127])
```
Comment (patrickvonplaten) [inline tests/pipelines/audio_diffusion/test_audio_diffusion.py:121]: This test should not change since we're using the original 1000 steps DDPM scheduler here.

**Review by patrickvonplaten — COMMENTED** [review]
Hey @dudulightricks,

Thanks for the PR. It is true that there is currently a problem when using DDPM with less than 1000 steps. However, this PR seems to fix multiple things - can we try to just fix the problem of `num_inference_steps` ? 

Also it looks like the current implementation differs from the old implementation even for `num_inference_steps=1000` which it shouldn't. Can we make sure that we don't do any unexpected changes? 

In short DDPM is supposed to be used for inference with 1000 steps - we need to make sure that this PR doesn't change the behavior for this case.

File: tests/pipelines/audio_diffusion/test_audio_diffusion.py (hunk 1 above)
Comment (dudulightricks) [reply tests/pipelines/audio_diffusion/test_audio_diffusion.py:121]: no we use 4 steps. please notice steps=4 in lines 109 and 114 before these rows.
I already checked nothing is changed when using 1000 as before (many test works with 1000 and not failing).

**Comment by dudulightricks** [comment]
@patrickvonplaten Fixed and Squashed.

1. I separated the fixes as you asked, and this PR only includes the infer steps.
2. I checked that it does not break anything for 1000 infer steps as before (also, there are many tests on 1000 steps which are not failing).

Thanks a lot for the review.

**Comment by dudulightricks** [comment]
@patrickvonplaten Rebased.

File: tests/pipelines/audio_diffusion/test_audio_diffusion.py (hunk 1 above)
Comment (patrickvonplaten) [reply tests/pipelines/audio_diffusion/test_audio_diffusion.py:121]: Thanks, yes you're 100% right.

File: tests/pipelines/ddpm/test_ddpm.py [hunk 2]
```diff
@@ -61,7 +61,7 @@ def test_inference(self):
 
         assert image.shape == (1, 32, 32, 3)
         expected_slice = np.array(
-            [5.589e-01, 7.089e-01, 2.632e-01, 6.841e-01, 1.000e-04, 9.999e-01, 1.973e-01, 1.000e-04, 8.010e-02]
+            [9.956e-01, 5.785e-01, 4.675e-01, 9.930e-01, 0.0, 1.000, 1.199e-03, 2.648e-04, 5.101e-04]
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply tests/pipelines/ddpm/test_ddpm.py:64]: So maybe just run just this test in a loop with:

```
for num_inference_steps in [2, 1000]:
    ...
```

Then I think we can merge this one :-)

[the remark this answers is not in the document]
Comment (dudulightricks) [reply tests/pipelines/ddpm/test_ddpm.py:63]: Done. I added it as 2 different tests since otherwise it reaches the timeout of 60 secs.

**Comment by dudulightricks** [comment]
@patrickvonplaten Fixed and Squashed.

**Comment by patrickvonplaten** [comment]
Cool, tested it and it works well - thanks for the fix :-)
